R/utils.R

```diff
@@ -170,3 +170,41 @@ check_connect_version <- function(using_version, minimum_tested_version = "1.8.8
   }
   invisible()
 }
+
+token_hex <- function(n) {
+  raw_to_hex <- function(raw) {
+    paste(as.character(raw), collapse = "")
+  }
+  raw <- as.raw(sample(0:255, n, replace = TRUE))
+  raw_to_hex(raw)
+}
+
+is_rendered <- function(app_mode) {
+  if ((app_mode) %in% c("rmd-static", "jupyter-static", "quarto-static")) {
+    TRUE
+  } else {
+    FALSE
+  }
+}
+
+is_interactive <- function(app_mode) {
+  interactive_app_modes <- c(
+    "shiny",
+    "rmd-shiny",
+    "jupyter-voila",
+    "python-api",
+    "python-dash",
+    "python-streamlit",
+    "python-bokeh",
+    "python-fastapi",
+    "python-shiny",
+    "quarto-shiny",
+    "tensorflow-saved-model",
+    "api"
+  )
+  if ((app_mode) %in% interactive_app_modes) {
+    TRUE
+  } else {
+    FALSE
+  }
+}
```

Comment:
Does the connect API know about the mapping between these two modes?

Reply:
The API doesn't contain data indicating this distinction, if that's what you mean. It feels wrong to have this information spread out like this. Definitely `rendered` vs `server-backed` isn't enough, because you also have static content that is neither. You could argue that we should have a property in the `content` API that indicates whether an app is rendered vs server-backed (i.e. restartable) vs neither.

A close analog to this is the `isWorkerApp` function used in the dashboard: https://github.com/rstudio/connect/blob/1d76aeddb339ec379614f747ee91761eca69b5fc/src/connect/worker/apptypes/apptypes.go#L158-L185

Reply:
Information on which apps are "worker" app modes _does_ exist in Go: https://github.com/rstudio/connect/blob/1d76aeddb339ec379614f747ee91761eca69b5fc/src/connect/worker/apptypes/apptypes.go#L158-L177

Reply:
We don't have any equivalent designation for "report" / "rendered" app modes, though, which feels like an oversight.

Reply:
Yeah, we don't have to (and probably shouldn't here), but it might be good to make an issue in the connect repo about what we might do in order to expose this better / more consistently?